src/table/BaseRecord.cc

```diff
@@ -90,6 +91,38 @@ void BaseRecord::assign(BaseRecord const& other, SchemaMapper const& mapper) {
     mapper.forEach(CopyValue(&other, this));  // use the functor we defined above
     this->_assign(other);                     // let derived classes assign their own stuff
 }
+
+namespace {
+
+class PrintField {
+public:
+
+    explicit PrintField(BaseRecord const * record, std::ostream * os) : _record(record), _os(os) {}
+
+    template <typename T>
+    void operator()(SchemaItem<T> const & item) const {
+        *_os << item.field.getName() << ": " << _record->get(item.key) << std::endl;
+    }
+
+private:
+    BaseRecord const * _record;
+    std::ostream * _os;
+};
+
+} // anonymous
+
+
+void BaseRecord::_stream(std::ostream & os) const {
+    PrintField functor(this, &os);
```

Comment:
Personally I would find this much clearer and easier to read if you wrote this as a lambda with a capture on this and os. It would put the function right next to where it was used, and reduce many lines in this file. This seems exactly to be the kind of thing that lambdas were made for. It would also negate the need to get the address of os to pass into the functor, just to keep the size of the functor small, you could simply reference capture os, and have it as a reference all the way down.

Also, if a functor is used, please consider reverse the arguments of BaseRecord and std::ostream to put them in the same order as operator<<. I dont know if it is standard to pass this first, and if so fine, but it seems weirdly asymmetric.